src/fibsem_tools/io/fibsem_h5.py

```diff
@@ -0,0 +1,314 @@
+    OFFSET,
+    MAGIC_NUMBER,
+    read_fibsem,
+)
+
+chunk_type = Union[Tuple[int, int], Tuple[int, int, int]]
+
+
+def create_fibsem_h5_dataset(
+    parent: Union[h5py.File, h5py.Group],
+    name: str,
+    data: FIBSEMData,
+    chunks: chunk_type = (128, 128),
+    **kwargs
+):
+    """
+    Create a HDF5 dataset from an existing FIBSEMData object.
+
+    :param parent: `h5py.File` or `h5py.Group` to place the dataset in
+    :param name: Name of the dataset
+    :param data: `FIBSEMData` instance
+    :param chunks: tuple describing the chunk size (Y, X)
+    :type chunks: tuple, optional
+    :param **kwargs: Keyword arguments are forwarded to `parent.create_dataset`
+    """
+
+    # If chunks has length 2, add a 1 for the ChanNum dimension
+    if len(chunks) == 2:
+        if data.shape[0] == data.header.ChanNum:
+            chunks = (1, *chunks)
+        elif data.shape[2] == data.header.ChanNum:
+            chunks = (*chunks, 1)
+
+    ds = parent.create_dataset(name, data.shape, data.dtype, chunks=chunks, **kwargs)
+    for key, value in data.header.__dict__.items():
+        ds.attrs[key] = value
+    ds[:] = data
+    return ds
+
+
+def create_fibsem_h5_dataset_external(
+    parent: Union[h5py.File, h5py.Group],
+    name: str,
+    data: FIBSEMData,
+    datfile: str,
+    **kwargs
+):
+    """
+    Create a HDF5 external dataset where the data is still stored in the dat file.
+    Note that the axes is rolled in the dat file versus FIBSEMData object
+
+    :param parent: `h5py.File` or `h5py.Group` to place the dataset in
+    :param name: Name of the dataset
+    :param data: `FIBSEMData` instance
+    :param **kwargs: Keyword arguments are forwarded to `parent.create_dataset`
+    """
+
+    # FIBSEMData is usually (ChanNum, YResolution, XResolution)
+    shape = (
+        data.header.YResolution,
+        data.header.XResolution,
+        data.header.ChanNum,
+    )
+    parent.create_dataset(
+        name,
+        data.shape,
+        ">i2",
+        external=[(datfile, 1024, h5py.h5f.UNLIMITED)],
+        **kwargs
+    )
+    for key, value in data.header.__dict__.items():
+        ds.attrs[key] = value
+    ds[:] = data
+    return ds
+
+
+def create_fibsem_h5_file(
+    h5_filename: str, dataset_name: str, data: FIBSEMData, force: bool = False, **kwargs
+):
+    """
+    Create HDF5 file with a single dataset
+
+    :param h5_filename: Name of the HDF5 file to create
+    :param dataset_name: Name of the dataset
+    :param data: `FIBSEMData` instance
+    :param force: Overrwrite existing HDF5 file if True. defaults to False.
+    :type force: bool, optional
+    :param **kwargs: Keyword arguments are forwarded to `parent.create_dataset`
+    """
+    # Use 'x' file permissions so we do not overwrite files
+    mode = "w" if force else "x"
+    f = h5py.File(h5_filename, mode)
+    create_fibsem_h5_dataset(f, dataset_name, data, **kwargs)
+    f.close()
+
+
+def _extract_raw_header(dat_filename: str):
```

Comment:
this should probably be moved into the .dat reader module

Reply:
moved in [cecf4d2](https://github.com/janelia-cosem/fibsem-tools/pull/18/commits/cecf4d299d3e2cc779115f1b0350f9b9f9e591e2)